### PR TITLE
Transition to ES6 modules.

### PR DESCRIPTION
**Pull request by trusktr**
WIP

### CODE REVIEW

**Comment by jasonLaster** [comment]
Thanks for helping us out @trusktr and sharing early. I personally welcome a new `commonjs` approach!

**Comment by trusktr** [comment]
@jasonLaster No prob, will handle unit tests once all is converted.

**Comment by trusktr** [comment]
@jasonLaster Would you guys prefer CommonJS or ES6 Modules?

**Comment by jasonLaster** [comment]
ES6 Modules. Up to you on the tests, although I think a proof of concept that we could merge would be great.

**Comment by trusktr** [comment]
Ok cool, I prefer ES6 Modules too, so I'll use those since it's the future. And yeah, I'd like to have it all working before merging. :D

**Comment by jasonLaster** [comment]
Would it be easy to follow https://github.com/babel/babel-library-boilerplate

@jmeas is mm core

**Comment by trusktr** [comment]
Awesome, thanks for sharing that.

File: src/dom-refresh.js [hunk 1]
```diff
@@ -1,36 +1,42 @@
...
 // -----------
 
-// Monitor a view's state, and after it has been rendered and shown
-// in the DOM, trigger a "dom:refresh" event every time it is
-// re-rendered.
+import _       from 'underscore';
+import helpers from './helpers';
 
-Marionette.MonitorDOMRefresh = function(view) {
+exports default {
```
Comment (jasonLaster) [inline src/dom-refresh.js]:
```js
var MonitorDomRefresh = function() {}
exports default MonitorDomRefresh;
```

File: src/error.js [hunk 2]
```diff
@@ -1,38 +1,46 @@
...
 // -----
 
+import _       from 'underscore';
+import helpers from './helpers';
+
 var errorProps = ['description', 'fileName', 'lineNumber', 'name', 'message', 'number'];
 
-Marionette.Error = Marionette.extend.call(Error, {
-  urlRoot: 'http://marionettejs.com/docs/v' + Marionette.VERSION + '/',
+var exports = {
```
Comment (jasonLaster) [inline src/error.js]:
```js
var Error = helpers.extend.call();
Error.extend = helpers.extend
export default Error
```

File: src/helpers.js [hunk 3]
```diff
@@ -147,3 +25,138 @@ var deprecate = Marionette.deprecate = function(message, test) {
 
 deprecate._warn = typeof console !== 'undefined' && (console.warn || console.log) || function() {};
 deprecate._cache = {};
+
+
+var exports = {
```
Comment (jasonLaster) [inline src/helpers.js]:
```js
var isNodeAttached = function() {};
var mergeOptions = function() {}

export default {
  isNodeAttached: isNodeAttached,
  mergeOptions: mergeOptions
}

```

File: src/trigger-method.js [hunk 4]
```diff
@@ -3,66 +3,73 @@
...
-  var splitter = /(^|:)(\w)/gi;
+import _ from 'underscore';
 
-  // take the event section ("section1:section2:section3")
-  // and turn it in to uppercase name
-  function getEventName(match, prefix, eventName) {
-    return eventName.toUpperCase();
-  }
+var exports = {
+  _triggerMethod: (function() {
```
Comment (jasonLaster) [inline src/trigger-method.js]: var _triggerMethod = function() {};

var triggerMethodOn = function() {}; 

export default {
  triggreMethodOn: triggerMethodOn
}

File: src/dom-refresh.js (hunk 1 above)
Comment (trusktr) [inline src/dom-refresh.js]: Oh wow. How'd I miss that. xD

File: src/helpers.js (hunk 3 above)
Comment (trusktr) [inline src/helpers.js]: Gotcha, I'll follow that pattern. :+1:

File: src/trigger-method.js (hunk 4 above)
Comment (trusktr) [inline src/trigger-method.js]: Got it, keep _methods private! :+1:

Comment (jasonLaster) [inline src/trigger-method.js]: Yea, small export statements. Just a style nit honestly.

Comment (trusktr) [inline src/trigger-method.js]: I noticed that _getValue in helpers.js is used outside of helpers.js.

Comment (trusktr) [inline src/trigger-method.js]: So is _triggerMethod. I'll just keep these in the exports for now, get it working, then we can think of what to do, if anything.

File: src/error.js (hunk 2 above)
Comment (trusktr) [inline src/error.js]: I'd love omitting semicolons.

File: src/error.js [hunk 5]
```diff
@@ -1,12 +1,15 @@
...
 // -----
 
+import _       from 'underscore';
+import helpers from './helpers';
+
 var errorProps = ['description', 'fileName', 'lineNumber', 'name', 'message', 'number'];
 
-Marionette.Error = Marionette.extend.call(Error, {
-  urlRoot: 'http://marionettejs.com/docs/v' + Marionette.VERSION + '/',
+var MarionetteError = helpers.extend.call(Error, {
```
Comment (trusktr) [inline src/error.js]: I renamed this to MarionetteError to avoid confusion. Comapre to having `var Error = ...;`.

File: src/features.js [hunk 6]
```diff
@@ -1,6 +1,11 @@
...
-Marionette.isEnabled = function(name) {
-  return !!Marionette.FEATURES[name];
+var FEATURES = {};
+
+function isEnabled(name) {
+  return !!FEATURES[name];
+}
+
+export default {
+  FEATURES: FEATURES,
```
Comment (trusktr) [inline src/features.js]: Is FEATURES used for anything?

Comment (jasonLaster) [inline src/features.js]: Not yet :)

**Comment by jasonLaster** [comment]
Heya @trusktr how's it going?

Just wanted to give you a heads up that were transitioning to a new branching scheme for a major release. Would you mind opening this PR again against next

**Comment by trusktr** [comment]
Yeah, sure!

**Comment by megawac** [comment]
:+1: can be shipped as a minor version I think. Can you add the relevant dependencies to the package.json and a `browserify: {
   transforms: [['babelify', {loose: true}]
}`

**Comment by trusktr** [comment]
Hey guys, due to a pivot moment at work and moving onto some new projects that won't be using Marionette for the time being, I might not have time to do this. I'll close, but come back later if time allows!

**Pull request closed without merging**